Verify GitHub evidences

damianmarti opened the pull request:
- Verify GitHub evidence when indexing events from Ponder.
- Check on GitHub repository contributor y languages, looking for the attested username and the skill attested.
- Added verifiedCount to developerSkills: this is the count of attestations for this skill with valid evidence.
- Added collaboratorCount to developerSkills: this is the count of attestations for this skill with valid evidence, AND the attester user has contributed to the project too.
- Basic score calculation: one point for the attestation, one point if this is verified, and one point if it's from a collaborator (we can improve it later).
- Added evidencesVerified and evidencesCollaborator to attestations: so we can show an evidence as verified or from a collaborator.
- Added background color based on evidence verification on UserAttestations.

With all this new information available from Ponder GraphQL, we can improve the data shown in the website.

GITHUB_TOKEN is not needed if the GitHub API endpoints are hit only a couple of times, but then it's rate-limited.

One future improvement is to avoid adding more score to a user if the same user attests to a user more than once. Issue created for this #30

damianmarti commented:
> Thx for tackling this @damianmarti !
> 
> Trying to check it, but I'm getting this when I run `yarn ponder:dev`, do I need to run some command previously to clean some stuff?
> 
> ```
> PS E:\GitHub\dev-creds> yarn ponder:dev
> 10:27:01 WARN  app        Local environment file (.env.local) not found
> 10:27:05 ERROR build      Error while executing 'src\Attestations.ts':
> AggregateError:
>     at internalConnectMultiple (node:net:1134:18)
>     at afterConnectMultiple (node:net:1715:7)
> 10:27:05 WARN  process    Failed intial build, starting shutdown sequence
> 10:27:05 FATAL process    Finished shutdown sequence, terminating (exit code 1)
> ```
> 
> Initially I was getting a similar message, but removed the `ponder-env.d.ts` file which was generated from the previous tests on a different branch, and it changed to the one I posted before.
> 
> Previous error:
> 
> ```
> 10:20:32 ERROR build      Error while executing 'src\Attestations.ts':
> Error: Error while executing src\Attestations.ts: Failed to load url redis (resolved id: redis) in E:/GitHub/dev-creds/packages/ponder/src/Attestations.ts. Does the file 
> exist?
>     at loadAndTransform (file:///E:/GitHub/dev-creds/packages/ponder/node_modules/vite/dist/node/chunks/dep-wTaJK0Jt.js:49073:21)
>     at ViteNodeServer._transformRequest (file:///E:/GitHub/dev-creds/packages/ponder/node_modules/vite-node/dist/server.mjs:393:16)
>     at ViteNodeServer._fetchModule (file:///E:/GitHub/dev-creds/packages/ponder/node_modules/vite-node/dist/server.mjs:359:17)
>     at ViteNodeRunner.directRequest (file:///E:/GitHub/dev-creds/packages/ponder/node_modules/vite-node/dist/client.mjs:250:46)
> 10:20:32 ERROR build      Error while executing 'src\Attestations.ts':
> Error: Error while executing src\Attestations.ts: Failed to load url redis (resolved id: redis) in E:/GitHub/dev-creds/packages/ponder/src/Attestations.ts. Does the file 
> exist?
>     at loadAndTransform (file:///E:/GitHub/dev-creds/packages/ponder/node_modules/vite/dist/node/chunks/dep-wTaJK0Jt.js:49073:21)
>     at ViteNodeServer._transformRequest (file:///E:/GitHub/dev-creds/packages/ponder/node_modules/vite-node/dist/server.mjs:393:16)
>     at ViteNodeServer._fetchModule (file:///E:/GitHub/dev-creds/packages/ponder/node_modules/vite-node/dist/server.mjs:359:17)
>     at ViteNodeRunner.directRequest (file:///E:/GitHub/dev-creds/packages/ponder/node_modules/vite-node/dist/client.mjs:250:46)
> ```

Sorry, you need to set REDIS_URL in .env.local (the same redis_url that you are using at the NextJS frontend).

Pabl0cks reviewed (commented):
> Sorry, you need to set REDIS_URL in .env.local (the same redis_url that you are using at the NextJS frontend).

Thx Damu! Working nicely now. Just a couple of nitpicks / questions:

- Should we normalize the evidence url in case some people don't add the `https://` or other edge cases?
- Was wondering if we should lower the bar for collaborator evidence, to don't need to match both language in the repo AND attested by another contributor. 

What I was lacking is I might need to attest SQL or Drizzle - Ponder from a collaborator, and it would get no verification badge right now since it will not be a language from the repo. So a badge that shows that it got verified but collaborator could be useful in those cases, we could show 2 badges per attestation, to show that is matching both.

damianmarti commented:
> > Sorry, you need to set REDIS_URL in .env.local (the same redis_url that you are using at the NextJS frontend).
> 
> Thx Damu! Working nicely now. Just a couple of nitpicks / questions:

Thanks @Pabl0cks for the review!

> 
> * Should we normalize the evidence url in case some people don't add the `https://` or other edge cases?

Added a check for GitHub URL with no https. I think it's enough for now. We should add validation on the frontend on the attestation form (there is already an open issue for that #4)

> * Was wondering if we should lower the bar for collaborator evidence, to don't need to match both language in the repo AND attested by another contributor.
> 
> What I was lacking is I might need to attest SQL or Drizzle - Ponder from a collaborator, and it would get no verification badge right now since it will not be a language from the repo. So a badge that shows that it got verified but collaborator could be useful in those cases, we could show 2 badges per attestation, to show that is matching both.

Makes sense. The languages from GitHub are a hard limitation. I changed this so the collaborator flag does not depend on languages.

Thanks!

bhavyagor12 reviewed (approved):
Thanks a ton @damianmarti this is working great, and looks GTM!!
Thanks @Pabl0cks for the review! already covered a bunch of great optimizations
